File: problems/239. Sliding Window Maximum/solution.cpp
class Solution
{
public:
    std::vector<int> maxSlidingWindow(std::vector<int> &nums, int k)
    {
        if (nums.size() == 0)
            return {};
        if (nums.size() == 1)
            return {nums[0]};
        int currMax;
        std::queue<int> q;
        for (int i = 0; i < k; i++)
        {
            currMax = std::max(currMax, nums[i]);
            q.push(nums[i]);
        }
        std::vector<int> result = {currMax};

        for (int i = k; i < nums.size(); i++)
        {
            int eject = q.front();
            q.pop();
            q.push(nums[i]);

            if (eject == currMax)
            {
                if (nums[i] >= currMax)
                    currMax = nums[i];
                else
                {
                    currMax = 0;
                    for (int j = i - k + 1; j < i + 1; j++)
                        currMax = std::max(currMax, nums[j]);
                }
            }
            else
            {
                currMax = std::max(currMax, nums[i]);
            }

            result.push_back(currMax);
        }

        return result;
    }
};

File: problems/239. Sliding Window Maximum/solution.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if not nums:
            return []
        queue = collections.deque(nums[:k])
        currMax = max(nums[:k])
        result = [currMax]
        for i in range(k, len(nums)):
            eject = queue.popleft()
            queue.append(nums[i])
            if eject == currMax:
                if (nums[i] >= currMax):
                    currMax = nums[i]
                else:
                    currMax = max(nums[i-k+1:i+1])
            else:
                currMax = max(currMax, nums[i])

            result.append(currMax)
        return result
